src/main/java/seedu/address/model/tag/UniqueTagList.java

```diff
@@ -0,0 +1,139 @@
+package seedu.address.model.tag;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.person.exceptions.DuplicatePersonException;
+import seedu.address.model.tag.exceptions.DuplicateTagException;
+import seedu.address.model.tag.exceptions.TagNotFoundException;
+
+/**
+ * A list of tags that enforces uniqueness between its elements and does not allow nulls.
+ * A tag is considered unique by comparing using {@code Tag#isSameTag(Tag)}. As such, adding and updating of
+ * tags uses Tag#isSameTag(Tag) for equality so as to ensure that the person being added or updated is
+ * unique in terms of identity in the UniqueTagList. However, the removal of a tag uses Tag#equals(Object) so
+ * as to ensure that the tag with exactly the same fields will be removed.
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Tag#isSameTag(Tag)
+ */
+public class UniqueTagList implements Iterable<Tag> {
+    private final ObservableList<Tag> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Tag> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains an equivalent tag as the given argument.
+     */
+    public boolean contains(Tag toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameTag);
+    }
+
+    /**
+     * Adds a tag  to the list.
+     * The tag must not already exist in the list.
+     */
+    public void add(Tag toAdd) {
```

Comment:
perhaps naming of the parameter can be improved such as `tagToBeAdded` for better readability

Reply:
Thanks for pointing it out.